### ISSUE TITLE
Support partial_merge and full_sorting_merge for pre-sorted tables using file/url/s3 engine

### ISSUE THREAD

**Issue by tolmalev** [ClickHouse/ClickHouse#58752]
> A clear and concise description of what is the intended usage scenario is.

full_sorting_merge is a very effective way to merge two large tables if it's known that tables are already sorted. 
But it currently only works for MergeTree* engines because only for these engines ClickHouse knowns that data is already sorted by some subset of columns. For all other engines clickhouse will have to sort tables in memory before applying merge join algorithm.
When working with external files like parquet it's possible that user knows that file is already ordered and it would be great to teach ClickHouse to use this information to optimise some operations and specifically joins.

> A clear and concise description of what you want to happen.

- 1st step: Let clickhouse know that file is sorted by some set of columns. Not sure what's the best way to pass this information from Query to Clickhouse. Some options: 
  - Add parameter to file/s3/url table function and table engine
  - Add setting engine_file_known_sort_order
  - Add new formats. e.g. SorterParquet, ...
  - For Parquet files it's possible to use `sorting_columns` field from RowGroup metadata. Although this is defined on row-group level, not on the file level. And to better use it, clickhouse should start writing this metadata field when creating parquet files.
- 2nd step: use information about ordering to enable full_sorting_merge and partial_merge without additional sorting step

**Describe alternatives you've considered**
> A clear and concise description of any alternative solutions or features you've considered.

I'm not aware of alternative ways to effectively use full_sorting_merge on external files.

**Additional context**

> Add any other context or screenshots about the feature request here.

**Comment by den-crane** [comment]
>full_sorting_merge is a very effective way to merge two large tables if it's known that tables are already sorted.

It sorts tables even with MergeTree, but sorting works faster because of shortcuts in sorting, pure merge join is not implemented

```sql
create table A ( A Int64, S String) Engine=MergeTree order by A as select number, repeat(' ', 200) from numbers(1e8);
set join_algorithm='full_sorting_merge';

select min(t.A), min(t.S),min(t.A), min(s.S) from A t join A s using (A);

1 row in set. Elapsed: 27.511 sec. Processed 200.00 million rows, 43.40 GB (7.27 million rows/s., 1.58 GB/s.)
Peak memory usage: 54.89 GiB.  ---<<<<<<-
```

**Comment by UnamedRus** [comment]
> For Parquet files it's possible to use sorting_columns field from RowGroup metadata. Although this is defined on row-group level, not on the file level. And to better use it, clickhouse should start writing this metadata field when creating parquet files.

Parquet files can have custom metadata on file level. (key_value_metadata)
![FileFormat](https://github.com/ClickHouse/ClickHouse/assets/9449405/5a7aabf0-d57d-4a26-a19d-f9827665813f)

Same about Arrow format

https://arrow.apache.org/docs/format/Columnar.html#custom-application-metadata

**Comment by tolmalev** [comment]
@UnamedRus that's true, but it won't be something standard and understood by all other libraries that work with Parquet.
But yeah, we could define some metadata field that will be recognised specifically by ClickHouse and allow it to apply optimisation for sorting

**Comment by tolmalev** [comment]
@den-crane 
Is it possible to apply same shortcuts that are used for MergeTree for pre-sorted external files?
